From 68becf58004a429be10d10b0df9307a32c46505c Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 23 Jul 2014 10:49:54 -0400
Subject: [virt] virtio_blk: verify if queue is broken after virtqueue_get_buf()

Message-id: <1406112611-9862-7-git-send-email-brueckner@redhat.com>
Patchwork-id: 86319
O-Subject: [RHEL7.1 PATCH 06/23] [virt] virtio_blk: verify if queue is broken after virtqueue_get_buf()
Bugzilla: 1088554
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
commit 7f03b17d5c3f3b63dcacf0e481cf94c9d07267ab
Author: Heinz Graalfs <graalfs@linux.vnet.ibm.com>
Commit: Rusty Russell <rusty@rustcorp.com.au>

    virtio_blk: verify if queue is broken after virtqueue_get_buf()

    In case virtqueue_get_buf() returned with a NULL pointer verify if the
    virtqueue is broken in order to leave while loop.

    Signed-off-by: Heinz Graalfs <graalfs@linux.vnet.ibm.com>
    Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

Bugzilla
--------
BZ 1088554
https://bugzilla.redhat.com/show_bug.cgi?id=1088554

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version v3.13-rc1
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=7f03b17d5c3f3b63dcacf0e481cf94c9d07267ab

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7725759

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/virtio_blk.c b/drivers/block/virtio_blk.c
index 1df5a36..f3cb1a2 100644
--- a/drivers/block/virtio_blk.c
+++ b/drivers/block/virtio_blk.c
@@ -292,6 +292,8 @@ static void virtblk_done(struct virtqueue *vq)
     req_done = true;
    }
   }
+  if (unlikely(virtqueue_is_broken(vq)))
+   break;
  } while (!virtqueue_enable_cb(vq));
  /* In case queue is stopped waiting for more buffers. */
  if (req_done)
-- 
1.7.1